### DISCUSSION TITLE
Music Assistant with SnapCast Client: No chunks available

### DISCUSSION THREAD

**Discussion by ab-tools** [music-assistant/support#3370] [Q&A] [answered]
Hello,

I'm trying to setup Music Assistant to play on a SnapCast client.

What I have done on the server:

- Installed Music Assistant server version 2.3.4 as Docker container using the `docker compose` running it network `host` mode exactly as recommended in the instructions here: https://music-assistant.io/installation/
- Added Snapcast as playback provider in Music Assistant.
- Connected both Spotify (with Client ID) as well as a local folder with one OGG file for testing.

What I have done on the client (Raspberry Pi Zero 2W):

- Installed latest Raspberry image (Debian 1:6.6.62-1+rpt1 (2024-11-25) aarch64 GNU/Linux)
- Installed PulseAudio using `apt`
- Installed latest SnapClient with Pulse from here: https://github.com/badaix/snapcast/releases/download/v0.29.0/snapclient_0.29.0-1_arm64_bookworm_with-pulse.deb
- Verified with `paplay` that I can successfully play an OGG audio file.

Result:

- The SnapClient shows up in Music Assistant as expected.
- But when I try to play any track on that player, it shows in Music Assistant as started for a fraction of a second and then immediately stops again.
- When I check the logs on the client with `journalctl -u snapclient` I see this:

```
Dec 26 10:31:16 micclient1 snapclient[1396]: Resolving host IP for: 192.168.16.48
Dec 26 10:31:16 micclient1 snapclient[1396]: Connecting to 192.168.16.48:1704
Dec 26 10:31:16 micclient1 snapclient[1396]: Connected to 192.168.16.48
Dec 26 10:31:16 micclient1 snapclient[1396]: My MAC: "e6:77:fd:b0:50:46", socket: 9
Dec 26 10:31:16 micclient1 snapclient[1396]: ServerSettings - buffer: 1000, latency: 0, volume: 25, muted: 0
Dec 26 10:31:16 micclient1 snapclient[1396]: Codec: flac, sampleformat: 48000:16:2
Dec 26 10:31:16 micclient1 snapclient[1396]: Resampling from 48000:16:2 to 44100:16:2
Dec 26 10:31:16 micclient1 snapclient[1396]: Player name: pulse, device: default, description: Let PulseAudio server choose the device>
Dec 26 10:31:16 micclient1 snapclient[1396]: Mixer mode: software, parameters: <none>
Dec 26 10:31:16 micclient1 snapclient[1396]: Sampleformat: 44100:16:0, stream: 44100:16:2
Dec 26 10:31:16 micclient1 snapclient[1396]: Setting property "application.icon_name" to "snapcast"
Dec 26 10:31:16 micclient1 snapclient[1396]: Setting property "media.role" to "music"
Dec 26 10:31:16 micclient1 snapclient[1396]: Using buffer_time: 100 ms, server: default
Dec 26 10:31:16 micclient1 snapclient[1396]: Start
Dec 26 10:31:16 micclient1 snapclient[1396]: Connecting to pulse
Dec 26 10:31:16 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:16 micclient1 snapclient[1396]: diff to server [ms]: -4.06362e+07
Dec 26 10:31:17 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:18 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:19 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:20 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:21 micclient1 snapclient[1396]: No chunks available
Dec 26 10:31:21 micclient1 snapclient[1396]: No chunk received for 5000ms, disconnecting from pulse.
Dec 26 10:31:21 micclient1 snapclient[1396]: Disconnecting from pulse
Dec 26 10:32:23 micclient1 snapclient[1396]: Stop
Dec 26 10:32:23 micclient1 snapclient[1396]: Disconnecting from pulse
Dec 26 10:32:23 micclient1 snapclient[1396]: Codec: flac, sampleformat: 48000:16:2
Dec 26 10:32:23 micclient1 snapclient[1396]: Resampling from 48000:16:2 to 44100:16:2
Dec 26 10:32:23 micclient1 snapclient[1396]: Player name: pulse, device: default, description: Let PulseAudio server choose the device>
Dec 26 10:32:23 micclient1 snapclient[1396]: Mixer mode: software, parameters: <none>
Dec 26 10:32:23 micclient1 snapclient[1396]: Sampleformat: 44100:16:0, stream: 44100:16:2
Dec 26 10:32:23 micclient1 snapclient[1396]: Setting property "application.icon_name" to "snapcast"
Dec 26 10:32:23 micclient1 snapclient[1396]: Setting property "media.role" to "music"
Dec 26 10:32:23 micclient1 snapclient[1396]: Using buffer_time: 100 ms, server: default
Dec 26 10:32:23 micclient1 snapclient[1396]: Start
Dec 26 10:32:23 micclient1 snapclient[1396]: Connecting to pulse
Dec 26 10:32:23 micclient1 snapclient[1396]: No chunks available
Dec 26 10:32:24 micclient1 snapclient[1396]: No chunks available
Dec 26 10:32:25 micclient1 snapclient[1396]: No chunks available
Dec 26 10:32:26 micclient1 snapclient[1396]: No chunks available
Dec 26 10:32:27 micclient1 snapclient[1396]: No chunks available
Dec 26 10:32:27 micclient1 snapclient[1396]: Stop
Dec 26 10:32:27 micclient1 snapclient[1396]: Disconnecting from pulse
```

Any idea what I might be missing here?

Best regards and thanks in advance for any tip on this
Andreas

**Comment by OzGav** [comment]
I haven't heard of anyone using this software. People just generally use a straight snapclient.

When I quickly googled PulseAudio I did see a report of issues connecting to the Snapserver so you might want to look into that. Additionally check if PulseAudio is installing the 0.27 or newer version

**Comment by ab-tools** [comment]
Hello OzGav,

first thanks a lot for your really quick reply, much appreciated!

You mean the SnapClient version?
I explicitly install version 0.29.0 (see the mentioned DEB download link above). Also `apt` confirms the installed version with:
```
snapclient is already the newest version (0.29.0-1).
```
As SnapClient is logging `No chunks available`, this indicated for me that it would not be receiving audio chunks from the server instead of an issue with writing audio to PulseAudio for output or you think different here?

Best regards and again, thanks for your support
Andreas

**Reply by OzGav** [reply]
Confirm that MA and your clients are on the same subnet/VLAN

**Reply by ab-tools** [reply]
Hello OzGav,

yes, that's confirmed:
This is a simple test setup here.

Just MA running on a server in the mentioned Docker container with network in `host` mode as recommended.
The server has the IP address `192.168.16.48`.

The Raspberry Pi client (which is the only SnapClient connected) is on the same wired network with IP address `192.168.16.131`.

I also do understand the log excerpt above in a way that they can at least talk to each other:
```
Dec 26 10:31:16 micclient1 snapclient[1396]: Resolving host IP for: 192.168.16.48
Dec 26 10:31:16 micclient1 snapclient[1396]: Connecting to 192.168.16.48:1704
Dec 26 10:31:16 micclient1 snapclient[1396]: Connected to 192.168.16.48
Dec 26 10:31:16 micclient1 snapclient[1396]: My MAC: "e6:77:fd:b0:50:46", socket: 9
```

Is there anything else I could try or check?

Best regards and thanks for your efforts
Andreas

**Reply by OzGav** [reply]
Are you using the MA inbuilt snapserver?

**Reply by ab-tools** [reply]
Yes, I do.

**Reply by OzGav** [reply]
@SantiagoSotoC

**Reply by DerPicknicker** [reply]
@ab-tools .. try higher buffer maybe bad connection?

**Reply by ab-tools** [reply]
First thanks for your reply!

I can't really imagine how the connection could be bad as this is a direct wired Ethernet connection (so no WiFi involved or so).
Also, from the logs it seems that the client does not get any chunk from the server.

But I can still try playing with the buffer size if you think this might help nonetheless:
Currently, this is all set to its defaults, what buffer and chunk size would you recommend trying?

Maybe @SantiagoSotoC has any further idea what could be the issue here?

Best regards and thanks again for all the support
Andreas

**Reply by DerPicknicker** [reply]
Hi @ab-tools ,

then the buffer sizes doesn't matter so your assumptions are totally right. 

Maybe try older snapclient version? I use till today v28 or v27. Because they're so reliable. This could help for debugging.

Best regards

Maybe @SantiagoSotoC has some more ideas.

**Comment by naund** [comment]
Why do you need pulseaudio? Usually pa is started after login. In your case pa has to be started as system wide daemon. But why not simply using alsa?

**Reply by ab-tools** [reply]
Hello naund,

thanks for your reply!

PulseAudio is required as I need to access the same audio output device also for Home Assistant audio announcements.
As I understand, this is not possible with ALSA, but it is with PulseAudio to have two processes accessing the same audio device. Please correct me if I'm wrong.

And yes, PulseAudio is running as system demon. I did follow this guide for proper installation:
https://github.com/FutureProofHomes/wyoming-enhancements/blob/master/snapcast/docs/2_install_pulseaudio.md

PulseAudio seems to be working fine, I can output audio with `paplay` without issues.

Just the SnapClient is not working right now for some reason with the log output as posted above.

Best regards and again, thanks for your support
Andreas

**Reply by OzGav** [reply]
Try playing to a browser. If that works then it is likely your pulseaudio RPi that is the problem. As suggested above try installing 0.28

**Comment by ab-tools** [comment]
Hello everyone,

first thanks again for all your support so far!

Looks like I should have just checked the Music Assistant logs to begin with as there are a lot of errors and warnings in there:
```
2024-12-30 14:44:40.610 INFO (MainThread) [music_assistant.streams] Start Queue Flow stream for Queue Snapweb client - crossfade: False
2024-12-30 14:44:40.819 ERROR (MainThread) [ffmpeg.1158] Stream error: Failed to stream track spotify://track:20I6sIOMTCkB6w7ryavxtO
2024-12-30 14:44:40.824 WARNING (MainThread) [music_assistant.audio.media_stream] Stream error on spotify--h4XuTKYb://track/20I6sIOMTCkB6w7ryavxtO
2024-12-30 14:44:41.039 ERROR (MainThread) [ffmpeg.1167] Stream error: Failed to stream track spotify://track:7Ar5Ovvs0Oq8zqqPBeG64g
2024-12-30 14:44:41.047 WARNING (MainThread) [asyncio] child process pid 1167 exit status already read:  will report returncode 255
2024-12-30 14:44:41.048 WARNING (MainThread) [music_assistant.audio.media_stream] Stream error on spotify--h4XuTKYb://track/7Ar5Ovvs0Oq8zqqPBeG64g
2024-12-30 14:44:41.054 ERROR (MainThread) [ffmpeg.1165] Stream error: Failed to stream track spotify://track:20I6sIOMTCkB6w7ryavxtO
2024-12-30 14:44:41.059 WARNING (MainThread) [music_assistant.audio] Could not determine integrated loudness of spotify--h4XuTKYb://track/20I6sIOMTCkB6w7ryavxtO - [in#0 @ 0xffff85e12b00] Error opening input: Invalid data found when processing input
Error opening input file -.
Error opening input files: Invalid data found when processing input
2024-12-30 14:44:41.259 ERROR (MainThread) [ffmpeg.1182] Stream error: Failed to stream track spotify://track:7Ar5Ovvs0Oq8zqqPBeG64g
2024-12-30 14:44:41.265 WARNING (MainThread) [music_assistant.audio] Could not determine integrated loudness of spotify--h4XuTKYb://track/7Ar5Ovvs0Oq8zqqPBeG64g - [in#0 @ 0xffffac412b00] Error opening input: Invalid data found when processing input
Error opening input file -.
Error opening input files: Invalid data found when processing input
2024-12-30 14:44:41.279 ERROR (MainThread) [ffmpeg.1184] Stream error: Failed to stream track spotify://track:6mmyRfUzsjrZ4WOdetPMRe
2024-12-30 14:44:41.287 WARNING (MainThread) [asyncio] child process pid 1184 exit status already read:  will report returncode 255
2024-12-30 14:44:41.288 WARNING (MainThread) [music_assistant.audio.media_stream] Stream error on spotify--h4XuTKYb://track/6mmyRfUzsjrZ4WOdetPMRe
2024-12-30 14:44:41.289 INFO (MainThread) [music_assistant.streams] Finished Queue Flow stream for Queue Snapweb client
2024-12-30 14:44:41.289 ERROR (MainThread) [ffmpeg.1157] Stream error: No audio data received from source
2024-12-30 14:44:41.504 ERROR (MainThread) [ffmpeg.1222] Stream error: Failed to stream track spotify://track:6mmyRfUzsjrZ4WOdetPMRe
2024-12-30 14:44:41.511 WARNING (MainThread) [music_assistant.audio] Could not determine integrated loudness of spotify--h4XuTKYb://track/6mmyRfUzsjrZ4WOdetPMRe - [in#0 @ 0xffffaa812b00] Error opening input: Invalid data found when processing input
Error opening input file -.
Error opening input files: Invalid data found when processing input
```

I searched for the error message and found this issue here:
https://github.com/music-assistant/hass-music-assistant/issues/2313#issuecomment-2395331001
But looks like this was resolved with 2.2.9 and I'm running 2.3.4 at the moment.

I have successfully authenticated to Spotify with an own Client ID.

What could cause the issues mentioned in the logs?

Best regards and thanks
Andreas

**Comment by ab-tools** [comment]
Hello,

first still a happy New Year to everyone! :-)

I would really appreciate if someone has an idea/tip on the error messages shown above.

@OzGav, did you see such an error before?

Best regards and thanks a lot in advance
Andreas

**Reply by OzGav** [reply]
Have you tried rolling your install back to 0.28 as suggested above? We need to eliminate all variables as no, we have not seen this error before but equally I haven’t heard of anyone using pulseaudio either. Also have you tried plying to a browser tab with snapweb?

**Reply by ab-tools** [reply]
Thanks for your quick reply again, OzGav.

I did try it with SnapCast web now and it doesn't work there either.
Same behavior: It seems to try starting a song for a brief moment, then it immediately stops again. Same error in the server log.

This seems completely unrelated to the client now, hence I did not try to rollback SnapClient.

As mentioned above, I did see similar error messages in logs in old issues of the HASS integration - I've linked one above.
Unfortunately, it looks like that all the issues of the old HASS integration are deleted/hidden now, so I can't access them anymore.

But from the error message in the server log this looks like it has issues streaming music from Spotify (unrelated to any client) - you agree with that?

Best regards
Andreas

**Reply by OzGav** [reply]
All issues remain on GH you just need to change the filter to showed the closed ones.

So if it isn’t Snapcast then you need to try a different player provider. Use a radio station if you have nothing else.

edit! Are you using your own clientid with Spotify?

**Reply by ab-tools** [reply]
Are you sure all issues of the HASS integration are still visible?

If I go to
https://github.com/music-assistant/hass-music-assistant/
I can't even see the "Issues" tab anymore...

And yes, as mentioned above, I do use my own ClientID.

I never tried to add anything else as Spotify is the main purpose, let me try a radio station right now.

**Reply by OzGav** [reply]
That is the old repo. We recently moved it to https://github.com/music-assistant/support

**Reply by ab-tools** [reply]
I see, then the issue I was referring to above with a similar error message is this one:
https://github.com/music-assistant/support/issues/2313#issuecomment-2395331001
But looks like this was resolved with 2.2.9 and I'm running 2.3.4 at the moment...

**Reply by ab-tools** [reply]
Thanks for the tip to test with the radio station, @OzGav, at least one step further now:
This works - both in Snap Web as well as the PulseAudio SnapClient 0.29.0. :-)

So, apparently, the issue had nothing to do with the client, but only with the connection to Spotify.

Any idea how to investigate this part further why the Spotify connection doesn't work for me?

**Reply by OzGav** [reply]
Are you using your own clientid with Spotify?

**Reply by ab-tools** [reply]
You asked that question a few times now, @OzGav. ;-)

Yes, I do use my own ClientId with Spotify.

Any idea what could cause this or what I should test?

Thanks again
Andreas

**Reply by OzGav** [reply]
We have seen before someone who thought they had successfully used a clientid but had not. Can you check the Spotify dashboard and look at the graphs to confirm there are api calls being made to your account.

**Reply by ab-tools** [reply]
Confirmed:
![image](https://github.com/user-attachments/assets/7e899589-c887-47c9-9807-0d3938231a58)

**Reply by OzGav** [reply]
Looking back at your original report you said this affected local files as well. Is that correct?

**Reply by ab-tools** [reply]
Honestly, I think that I simply did (and still do :-) ) not understand how this local file storage works:
I do not plan to use local files at the end, so I just copied a single OGG file with a song to the `media` directory.

And yes, this directory is also exposed to the Docker container.

I originally thought that it would just use the local file instead of Spotify as source if it's the same song, but honestly, not sure if that worked.

And when I check the local file option in Music Assistant this does not show up:
![image](https://github.com/user-attachments/assets/5229199f-a431-44a0-a363-0a024e083b6e)

So I do think the issue really only occurs with Spotify as source.

**Reply by OzGav** [reply]
It would be good for you to confirm. You need a single file with some basic tags. If you add that to a directory which is structured ARTIST/ALBUM that is best. Make sure the folder name exactly matches the tagged names.

You should then find the track appear in the MA library after a sync. Trying playing that and confirm it plays without issue. I agree this is most likely Spotify but if it isn’t too much trouble it is handy to rule everything out.

Ultimately, this appears to be a problem in some way with your install as I have seen no reports with this exact error before. There are some current Spotify issues so the solution to one of those may have a flow on effect for you. For example, https://github.com/music-assistant/support/issues/3222

**Reply by ab-tools** [reply]
I don't have any music as files here and if I download something from e. g. YouTube, it won't have any tags, I guess?

Do you have a "test file" that I can use which includes the required tags to make it work correctly?
Happy to give it a try then.

**Reply by OzGav** [reply]
Ok if you literally don’t have any local music then you should remove the file system providers

**Reply by ab-tools** [reply]
Yeah, I don't have any local music at all - I just added it for testing, nothing else. Will remove it again.

To confirm that the general setup (also with PulseAudio SnapClient) is working fine, I successfully tested with a radio station and this is working correctly as mentioned above.

If you have any further idea what I could test/try regarding the Spotify issue, please let me know.

And thanks again for all your support.

**Reply by OzGav** [reply]
The next thing is to work out what does and doesn’t work with Spotify. So if you can trying playing an Artist, then album, then track, then playlist and see if you can identify any common factor… or does nothing work?

**Reply by OzGav** [reply]
If nothing works then look at your network for something blocking Spotify

**Comment by ab-tools** [comment, marked as answer]
@OzGav, I'm sorry, found the issue and this is totally my fault...

Few months ago I started testing with Music Assistant and just for this purpose I subscribed to Spotify (as I read in your documentation that it would only work with a subscription, not with free accounts). That was one of the Spotify test subscriptions they offered.
Apparently, this expired by now without me noticing it.

I just noticed it now when I logged in to Spotify directly to check API usage once again.

After I extended the Spotify Premium subscription all is working fine, so likely the only issue was that there was no paid subscription active at Spotify!

Really sorry that I wasted your time with that, should have checked that before.

But to conclude if someone else is reading this thread:

1. Using SnapClient with PulseAudio is no problem, tested with version 0.29.0.
2. REALLY make sure you are using a paid Spotify subscription!

I'm wondering, @OzGav, can you detect using the Spotify API if it is a paid subscription or not?
If so, it would be really helpful if you could write a BIG warning message into the Music Assistant logfile that an unsupported free Spotify account is used.

Currently, it just shows success messages about the account login in the logs like this one:
`
2024-12-26 09:31:14.853 INFO (MainThread) [music_assistant.spotify] Successfully logged in to Spotify as Andreas
`
This way it did not come to my mind that actually the Spotify account is the issue. If you could check and warn in case of free accounts that would really be very helpful.

In any case, thanks again for all your support, that is really much appreciated
Andreas

**Reply by OzGav** [reply]
No worries. Normally you do get a “Only premium accounts are supported” log entry but your expired premium account may be an edge case.

**Reply by ab-tools** [reply]
Just double-checked again and there was for sure no such log entry - maybe that is something that could be improved to also detect expired premium accounts like mine.

**Reply by OzGav** [reply]
I’ll ask the team

**Reply by ab-tools** [reply]
Thanks!
